### PR TITLE
[SPARK-44082][SQL] Generate operator does not update reference set properly

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Before
```
== Optimized Logical Plan ==
Project [col1#2, col2#19](https://issues.apache.org/jira/browse/SPARK-44082#2,%20col2#19)
+- Generate replicaterows(sum#17L, col1#2, col2#3), [2], false, [col1#2, col2#3](https://issues.apache.org/jira/browse/SPARK-44082#2,%20col2#3)
+- Filter (isnotnull(sum#17L) AND (sum#17L > 0))
+- Aggregate [col1#2, col2#19](https://issues.apache.org/jira/browse/SPARK-44082#2,%20col2#19), [col1#2, col2#19, sum(vcol#14L) AS sum#17L](https://issues.apache.org/jira/browse/SPARK-44082#2,%20col2#19,%20sum(vcol#14L)%20AS%20sum#17L)
+- Union false, false
:- Aggregate [col1#2](https://issues.apache.org/jira/browse/SPARK-44082#2), [1 AS vcol#14L, col1#2, first(col2#3, false) AS col2#19](https://issues.apache.org/jira/browse/SPARK-44082#14L,%20col1#2,%20first(col2#3,%20false)%20AS%20col2#19)
: +- LogicalRDD [col1#2, col2#3](https://issues.apache.org/jira/browse/SPARK-44082#2,%20col2#3), false
+- Project [-1 AS vcol#15L, col1#8, col2#9](https://issues.apache.org/jira/browse/SPARK-44082#15L,%20col1#8,%20col2#9)
+- LogicalRDD [col1#8, col2#9](https://issues.apache.org/jira/browse/SPARK-44082#8,%20col2#9), false
```

During the execution `Generate replicaterows(sum#17L, col1#2, col2#3), [2], false, [col1#2, col2#3]` couldn't find `col2#3` in `[col1#2,col2#19,sum#17L]`. So we need generate to update the reference attributes properly.

after this fix the plan is correct now
```
== Optimized Logical Plan ==
Project [col1#2, col2#19](https://issues.apache.org/jira/browse/SPARK-44082#2,%20col2#19)
+- Generate replicaterows(sum#17L, col1#2, col2#19), [2], false, [col1#2, col2#19](https://issues.apache.org/jira/browse/SPARK-44082#2,%20col2#19)
+- Filter (isnotnull(sum#17L) AND (sum#17L > 0))
+- Aggregate [col1#2, col2#19](https://issues.apache.org/jira/browse/SPARK-44082#2,%20col2#19), [col1#2, col2#19, sum(vcol#14L) AS sum#17L](https://issues.apache.org/jira/browse/SPARK-44082#2,%20col2#19,%20sum(vcol#14L)%20AS%20sum#17L)
+- Union false, false
:- Aggregate [col1#2](https://issues.apache.org/jira/browse/SPARK-44082#2), [1 AS vcol#14L, col1#2, first(col2#3, false) AS col2#19](https://issues.apache.org/jira/browse/SPARK-44082#14L,%20col1#2,%20first(col2#3,%20false)%20AS%20col2#19)
: +- LogicalRDD [col1#2, col2#3](https://issues.apache.org/jira/browse/SPARK-44082#2,%20col2#3), false
+- Project [-1 AS vcol#15L, col1#8, col2#9](https://issues.apache.org/jira/browse/SPARK-44082#15L,%20col1#8,%20col2#9)
+- LogicalRDD [col1#8, col2#9](https://issues.apache.org/jira/browse/SPARK-44082#8,%20col2#9), false
```

### Why are the changes needed?

Bug fix

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
@cloud-fan

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala [hunk 1]
```diff
@@ -3613,6 +3613,21 @@ class DataFrameSuite extends QueryTest
...
+    val schema = StructType(Array(
+      StructField("col1", StringType, true),
+      StructField("col2", StringType, true)
+    ))
+    val rowData = Seq(Row("100", "20000"),
+      Row("100", "100000"),
+      Row("300", "3000"))
+    val df1 = spark.createDataFrame(spark.sparkContext.parallelize(rowData), schema)
+    val df2 = spark.createDataFrame(spark.sparkContext.parallelize(rowData), schema)
+    df1.dropDuplicates(Seq("col1")).exceptAll(df2).collect.foreach { row =>
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:3627]: How is this test case related to the `Generate` operator?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala [hunk 2]
```diff
@@ -296,6 +296,8 @@ case class Generate(
 
   override protected def withNewChildInternal(newChild: LogicalPlan): Generate =
     copy(child = newChild)
+
+  override lazy val references: AttributeSet = generator.references
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:300]: The default implementation is kind of `generator.references -- producedAttributes`. Ideally these two should have no conjunction, and is the same as `generator.references`.

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala (hunk 1 above)
Comment (amaliujia) [reply sql/core/src/test/scala/org/apache/spark/sql/DataFrameSuite.scala:3627]: This is not because of the `Generate` operator but because of `ReplaceDeduplicateWithAggregate` rule which replace duplicate with aggregate which contains columns that current Generator operator does not know.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala (hunk 2 above)
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala:300]: so if it is valid bug, we should target at fixing `producedAttributes`?
